include/deal.II/grid/connectivity.h

```diff
@@ -0,0 +1,1477 @@
+          return line_vertices;
+
+        Assert(false, ExcNotImplemented());
+
+        return cell_entities;
+      }
+
+    private:
+      const unsigned int        dim;
+      std::vector<unsigned int> cell_types;
+
+      CRS<T> line_vertices;
+
+      std::vector<unsigned char> line_orientation;
+
+      CRS<T> quad_vertices;
+      CRS<T> quad_lines;
+
+      std::vector<unsigned char> quad_orientation;
+
+      CRS<T> cell_entities;
+      CRS<T> neighbors;
+
+      std::vector<unsigned int> quad_types;
+    };
+
+
+
+    /**
+     * Create the transposed of a compressed row storage matrix.
+     */
+    template <typename T>
+    CRS<T>
+    transpose(const CRS<T> &con)
+    {
+      auto        con_copy = con;
+      auto &      col      = con_copy.col;
+      const auto &ptr      = con_copy.ptr;
+
+      CRS<T> con_t;
+      auto & col_t = con_t.col;
+      auto & ptr_t = con_t.ptr;
+
+      for (unsigned int i = 0; i < ptr.size() - 1; i++)
+        std::sort(col.data() + ptr[i], col.data() + ptr[i + 1]);
+
+      col_t.clear();
+      ptr_t.clear();
+
+      col_t.reserve(col.size());
+      ptr_t.reserve(
+        col.size() == 0 ? 0 : *std::max_element(col.begin(), col.end()));
+
+      std::vector<std::pair<T, T>> temp;
+      temp.reserve(col.size());
+
+      for (unsigned int i = 0; i < ptr.size() - 1; i++)
+        for (std::size_t j = ptr[i]; j < ptr[i + 1]; j++)
+          temp.emplace_back(col[j], i);
+
+      std::sort(temp.begin(), temp.end());
+
+      T r = -1;
+
+      for (unsigned int i = 0; i < temp.size(); i++)
+        {
+          if (r != temp[i].first)
+            {
+              r = temp[i].first;
+              ptr_t.push_back(col_t.size());
+            }
+
+          col_t.push_back(temp[i].second);
+        }
+
+      ptr_t.push_back(col_t.size());
+
+      return con_t;
+    }
+
+
+
+    /**
+     * Determine the neighbors of a cell by visiting its faces and looking for
+     * a cell which is not equal to this cell.
+     */
+    template <typename T>
+    CRS<T>
+    determine_neighbors(const CRS<T> &con_cf, const CRS<T> &con_fc)
+    {
+      const auto &col_cf = con_cf.col;
+      const auto &ptr_cf = con_cf.ptr;
+      const auto &col_fc = con_fc.col;
+      const auto &ptr_fc = con_fc.ptr;
+
+      CRS<T> con_cc;
+      auto & col_cc = con_cc.col;
+      auto & ptr_cc = con_cc.ptr;
+
+      col_cc.clear();
+      ptr_cc.clear();
+
+      ptr_cc.push_back(0);
+
+      for (unsigned int i_0 = 0; i_0 < ptr_cf.size() - 1; i_0++)
+        {
+          for (std::size_t j_0 = ptr_cf[i_0]; j_0 < ptr_cf[i_0 + 1]; j_0++)
+            {
+              T temp = -1;
+
+              for (std::size_t j_1 = ptr_fc[col_cf[j_0]];
+                   j_1 < ptr_fc[col_cf[j_0] + 1];
+                   j_1++)
+                if (i_0 != col_fc[j_1])
+                  temp = col_fc[j_1];
+
+              col_cc.emplace_back(temp);
+            }
+
+          ptr_cc.push_back(col_cc.size());
```

Comment:
Do we know that this search always succeeds or do we allow to enter `-1`? Add either an assertion or document the exceptional case.

Reply:
I have documented that `-1` means boundary face.